open-apparel-registry/open-apparel-registry
# Polygonal search

## Overview

Allows users to filter facilities search to within a polygonal boundary. 

Users are now able to draw a polygon on the map. The facility search
results are filtered to only include the facilities which are contained
within the boundaries of the polygon.

The polygon is rendered based on the boundaries filter, so it renders
correctly on page refreshes based on the query string in the URL. The
polygon is also removed when the user resets the filters, thereby
removing the boundaries.

Redux functionality has been added to support the polygonal search
feature. Given a list of points in the form of `[[lat, long]]` in the
redux state, a query string is created when searching for facilities
which filters the facilities search to only return facilities within
the polygon described by the given points.

In addition, the points listed in the URL query string are able to be
parsed in order to rehydrate the redux state on load.

The API facilities search now takes a query string in the form
`&boundaries=lat,lng&boundaries=lat,lng` etc, where each incidence of
`&boundaries=lat,lng` represents a single point of a polygon. Search
results are filtered to only include facilities for which their
location is contained within the polygon.

Connects #793 

## Demo

![polygon](https://user-images.githubusercontent.com/21046714/75701102-246ed000-5c81-11ea-9038-c70a614bc68c.gif)

## Notes

I render the polygon as an SVG to avoid blocking clicks on the main map.

## Testing Instructions

* Check out this branch
* Run `vagrant ssh` and `./scripts/server`
* Click on the polygon button on the map, then click on points until you've drawn a polygon.
* The search results should be filtered to only show results within the polygon.
* You should be able to click on search results as usual.
* Refresh the page; the polygon should remain unchanged.
* Click the trashcan button and the polygon filter should be removed and the search re-run.
* Create a new polygon. 
* Click 'reset' in the search tab.
* The polygon filter should be removed. 

## Checklist
